GitHub_LearningTests/Feature/Login/Presentation/LoginPresenterTests.swift

```diff
@@ -0,0 +1,328 @@
+    }
+    
+    func test_givenUnvalidEmail_whenLogIn_thenInvalidEmailAlertIsThrown() {
+        // given
+        let testEmail = "test@gmail.c"
+        let testPassword = "Test123567"
+        let alertTitle = "Invalid Email"
+        let builder = LoginBuilderDummy()
+        let provider = AlertProviderControllerDummy()
+        setUpTests(loginBuilder: builder, alertProvider: provider)
+        
+        // when
+        builder.email = testEmail
+        builder.password = testPassword
+        sut.logIn()
+        
+        // then
+        XCTAssertNotNil(provider.alert)
+        XCTAssert(alertTitle == provider.alert?.title)
+    }
+    
+    func test_givenUnvalidPassword_whenLogIn_thenInvalidPasswordAlertIsThrown() {
+        // given
+        let testEmail = "test@gmail.com"
+        let testPassword = "Test"
+        let alertTitle = "Invalid Password"
+        let builder = LoginBuilderDummy()
+        let provider = AlertProviderControllerDummy()
+        setUpTests(loginBuilder: builder, alertProvider: provider)
+        
+        // when
+        builder.email = testEmail
+        builder.password = testPassword
+        sut.logIn()
+        
+        // then
+        XCTAssertNotNil(provider.alert)
+        XCTAssert(alertTitle == provider.alert?.title)
+    }
+    
+    func test_givenUnvalidPassword_whenLogIn_thenLogInWithIsNotCalled() {
+        // given
+        let testEmail = "test@gmail.com"
+        let testPassword = "Test"
+        let loginController = LoginControllerDummy()
+        let builder = LoginBuilderDummy()
+        setUpTests(loginBuilder: builder, loginController: loginController)
+        
+        // when
+        builder.email = testEmail
+        builder.password = testPassword
+        sut.logIn()
+        
+        // then
+        XCTAssertNil(loginController.isLogedIn)
+    }
+    
+    private class LoginBuilderDummy: LoginBuilder {
+        var email: String?
+        var password: String?
+        
+        var isMandatoryDataFilled: Bool {
+            guard let email = email, let password = password else { return false }
+            return !email.isEmpty && !password.isEmpty
+        }
+        
+        var isEmailValid: Bool {
+            guard let email = email else { return false }
+            return email.isEmail
+        }
+        
+        var isPasswordValid: Bool {
+            guard let password = password else { return false }
+            return password.containsNumber && password.containsCapital && password.length > 7
+        }
+        
+        func build() throws -> LoginCredentials {
+            guard let email = email, let password = password else { throw LoginBuilderError.missingMandatoryData }
+            guard isEmailValid else { throw LoginBuilderError.invalidEmail }
+            guard isPasswordValid else { throw LoginBuilderError.invalidPassword }
+            return LoginCredentials(email: email, password: password)
+        }
```

Comment:
Tady predpokladam, ze vubec nechces resit builder, ale to jestli presenter spravne nastavuje proprety a jestli spravne reaguje na property, pripadne build metodu. Proto neni treba kopirovat implementaci builderu, ale v initu mu dej co ma v techto pripadech vracet. Tj budes mit init, kde mu reknes co ma tato instance vratit pro `isMandatoryDataFilled`, `isEmailValid`, `isPasswordValid` a `build()`.